Enhance Currency Exchange Management

tundebabzy opened the pull request:
There was a discussion in #9973 about managing exchange rates.

My conclusion was to create a Settings doctype for Currency Exchange where the behaviour for fetching exchange rates can be tweaked. Presently, ERPNext cannot help users enforce IAS 21 standards of choosing exchange rates which recommends that the exchange rate used should be the one in effect on the transaction date. According to IAS 21 standards, exchange rates become stale at the end of the day it takes effect. With this new setting, this can be enforced.

![screenshot from 2017-08-11 01-50-31](https://user-images.githubusercontent.com/818803/29197950-d725148c-7e37-11e7-8b53-8cc96984ed3d.png)

![screenshot from 2017-08-11 01-54-06](https://user-images.githubusercontent.com/818803/29198015-325b54ba-7e38-11e7-985f-a12d387c5db4.png)

"Allow Stale Exchange Rates" will be the default and doesn't change ERPNext's present behaviour. If unchecked, the user can then specify how many days to consider as stale (1 by default).

With this change, we can now add something like this to the `get_exchange_rate` function:
```(python)
stale_date = None
if not Currency_Settings.allow_stale:
    stale_date = get_datetime_str(add_days(transaction_date, days=-Currency_Settings.stale_days))
filters = [
    ["from_currency", "=", from_currency],
    ["to_currency", "=", to_currency]
]
if stale_date:
    filters.append(["date", ">", stale_date])
else:
    filters.append(["date", "<=", get_datetime_str(transaction_date)])

...
if not exchange_rates:
    frappe.msgprint("Please save the current exchange rate in Currency Exchange")
...
```

bcornwellmott committed:
Added Quote Status to RFQ Suppliers with No Quote

bcornwellmott committed:
Minor cleanup

bcornwellmott committed:
Codacy fixes

bcornwellmott committed:
Added tests

bcornwellmott committed:
Fixed codacy issues

rohitwaghchaure committed:
Set write permission to sales manger for permlevel 1 in Quotation doctype

bcornwellmott committed:
More codacy issues

bcornwellmott committed:
MORE CODACY

rohitwaghchaure committed:
[Fix] Auto batch creation not working for purchase invoice is update stock is enabled (#10348)

rohitwaghchaure committed:
[Fix] Opportunity status has not changed after making of quotation against it if items has not added in the opportunity (#10351)

mbauskar committed:
[hotfix] remove the white spaces from condition and formula fields (#10331)

* [hotfix] remove the white spaces from condition and formula fields

* [tests] added tests cases for whitespaces in formula and condition fields

rohitwaghchaure committed:
[Fix] Getting an error cost center is not defined in the purchase receipt, even if it's defined in the company (#10320)

mbauskar commented:
@tundebabzy,

Can you please check and fix the codecy issues?, Closing the PR for now, Please reopen the PR once fixed.

rmehta commented on erpnext/setup/doctype/currency_exchange_settings/test_currency_exchange_settings.py:
```diff
@@ -0,0 +1,30 @@
+
+class TestCurrencyExchangeSettings(unittest.TestCase):
+	def test_change_setting(self):
+		settings_doc = frappe.get_doc("Currency Exchange Settings", "Currency Exchange Settings")
```
@tundebabzy such tests are not needed. You should only write tests to test code in your own controller.

Also this can also be part of "Accounts Settings"

tundebabzy committed:
refactor `get_exchange_rate`

tundebabzy committed:
make item name read only and add hide if not saved (#10382)

shreyashah115 committed:
[UI test] Buying module - Get Supplier Quotations (#10375)

* Get quotations for Suppliers

* Calculate taxes and charges for Supplier quotations

* Added paths for tests

* Codacy fix

* Improvised the page route function

mbauskar committed:
bumped to version 8.8.2

Commit:
send email via the background jobs (#10374)

vishdha committed:
[UI Test] sales invoice (#10394)

* [UI Test] Sales Invoice

* [UI Test] Sales Invoice

mbauskar committed:
bumped to version 8.8.3

mbauskar committed:
[docs] Translate the Schools Assessment section documentation to Spanish (#10398)

* [docs] Translate the Schools app documentation to Spanish

* [docs] Translate the Schools Student section documentation to Spanish

* [docs] Translate the Schools SETUP section documentation to Spanish

* [docs] Translate the Schools ADMISSION section documentation to Spanish

* [docs] Translate the Schools FEES section documentation to Spanish

* [docs] Translate the Schools SCHEDULE section documentation to Spanish

* [docs] Translate the Schools Assessment section documentation to Spanish

tundebabzy committed:
disables conversion rate field if stale rates not allowed

shreyashah115 committed:
[UI Tests] Buying Module - Supplier quotations and Purchase orders (#10397)

* Get Supplier Quotation with item wise discount

* Get purchase orders

* Added paths

* Codacy fix

tundebabzy committed:
more test cases

more test case...

test `get_exchange_rate` behaviour with stale not allowed in sett..

fix currency exchange test case

tundebabzy committed:
properly keep track of document currency and company currency

tundebabzy commented:
TASKS:
- [x] Currency Exchange Settings Doctype
- [x] validation
- [x] In `get_exchange_rate` check if settings allows for stale exchange rate
- [x] In forms, make conversion rate field read only if settings does not allow stale exchange rates
- [ ] Add documentation

Currency Settings fields:
1. Allow Stale Exchange Rate
2. Stale exchange rate days

tundebabzy commented:
Please review but don't merge yet. #10410 first has to be merged and #10145 needs to be fixed in order to proceed.

shreyashah115 committed:
[UI Test] Buying module - Multiple tests for Purchase orders (#10414)

* Extended timeout to avoid rare failures

* Get purchase orders with discount on grand total

* Get purchase orders with discount on individual items

* Get purchase orders and calculate taxes

* Added paths

utkarsh-goswami committed:
[UI Test Salary] To test the salary structure and the salary slip (#10407)

* Test for salary structure and salary slip

* Test for salary structure and salary slip

* Test for salary structure and salary slip

* Travis CI fixes

* Updated

* Updated

* travis fixes

* Travis fixes

Commit:
new item group filter in the stock ledger report (#10400)

tundebabzy committed:
update multi currency accounting documentation to remove incorrect info (#10404)

tundebabzy committed:
Transaction Forms (Payment Entry, Sales Invoice, Purchase Invoice) can forget company currency (#10409) (#10410)

* properly keep track of document currency and company currency

* Update taxes_and_totals.js

tundebabzy committed:
remove empty first row if needed (#10378)

refactor code: move global functions to callback scope

Zlash65 committed:
travis failing issue fix (#10420)

nabinhait committed:
[fix] Update monthly sales value on Company (#10417)

* [fix] Update monthly sales value on cancellation of invoice, also update modified datestamp

* Company monthly sales: patches and test cases

vishdha committed:
[UI Test] Purchase Invoice (#10422)

* [UI Test] Purchase Invoice

* [fix] Travis failed

shreyashah115 committed:
[UI Test] Buying - Create Purchase receipt (#10421)

* Create purchase receipt after submitting the purchase order

* Checks quantity and rate in stock ledger

Zlash65 committed:
made test less dependent on python test data (#10424)

mbauskar committed:
[hotfix] fixes for https://github.com/frappe/erpnext/issues/9782 calculate leave without pay before subtracting holidays (#10355)

rohitwaghchaure committed:
[Fix] Due date not fetching on the payment entry against the invoices (#10419)

mbauskar committed:
[hotfix] fixed type in gst_state_code_data.json file (#10437)

tundebabzy committed:
ui test for sales invoice

Zlash65 committed:
path to test updated (#10438)

vishdha committed:
[UI Test] UI Test for Purchase Receipt Added (#10434)

utkarsh-goswami committed:
[UI Test Recruitment] To test Recruitment in HR module (#10441)

* Tests for Recruitment

* Codacy fix

utkarsh-goswami committed:
Test for Process Payroll (#10447)

vishdha committed:
[UI Test] UI Test added for Stock Reconciliation (#10432)

* [UI Test] UI Test added for Stock Reconciliation

* [fix] Codacy

vishdha committed:
[UI Test] travis.yml modified for fresh UI Test (#10446)

* [UI Test] travis.yml modified for fresh UI Test

* [fix] travis

* Update test_production_order.js

* Update test_production_order.js

* Update test_purchase_order_receipt.js

* Update tests.txt

* Update test_purchase_order_receipt.js

KanchanChauhan committed:
[Minor] Title added to rfq web view (#10454)

tundebabzy committed:
add new settings in Accouts Settings

tundebabzy committed:
adds validation

tundebabzy committed:
tests validation

mbauskar committed:
[hotfix] fixed mysql syntax error in Student and Guardian Contact Details report (#10450)

mbauskar committed:
[hotfix] fixed the mysql syntax error issue on stock ledger report (#10442)

rohitwaghchaure committed:
[Fix] Tax rule is not working for the parent customer groups (#10458)

rohitwaghchaure committed:
[Fix] Making of production order from so, system not fetched the items from the sales order item if packing list has items and vice versa (#10453)

nabinhait committed:
Tax rate in Itemised sales/purchase register and cleanups (#10439)

Zlash65 committed:
[ui test ] Travis failing fix  (#10468)

* Value in assertion was wrong

* Update test_purchase_order_receipt.js

nabinhait committed:
Additional GST related fields in Invoice and Sales/Purchase Register reort (#10405)

mbauskar committed:
bumped to version 8.8.4

rmehta committed:
[docs] for print style (#10452)

codingCoffee committed:
converted production tests to INR from USD (#10472)

tundebabzy committed:
patch for new settings
